libautoscoper/src/PSO.cpp

```diff
@@ -66,55 +86,52 @@ void pso(float *positions, float *velocities, float *pBests, float *gBest, unsig
       do_this = false;
     }
 
-    float currentBest = host_fitness_function(gBest);
+    Particle* currentBest = gBest;
 
-    for (int i = 0; i < NUM_OF_PARTICLES*NUM_OF_DIMENSIONS; i++)
+    // Update the velocities and positions
+    for (int i = 0; i < NUM_OF_PARTICLES; i++)
     {
-      float rp = getRandomClamped();
-      float rg = getRandomClamped();
+      for (int j = 0; j < NUM_OF_DIMENSIONS; j++) {
+        float rp = getRandomClamped();
+        float rg = getRandomClamped();
 
-      velocities[i] = OMEGA * velocities[i] + c1 * rp*(pBests[i] - positions[i]) + c2 * rg*(gBest[i%NUM_OF_DIMENSIONS] - positions[i]);
-
-      positions[i] += velocities[i];
+        velocities.at(i*NUM_OF_DIMENSIONS + j) = OMEGA * velocities.at(i*NUM_OF_DIMENSIONS + j) + c1 * rp*(pBest.at(i)->position.at(j) - particles->at(i)->position.at(j)) + c2 * rg * (gBest->position.at(j) - particles->at(i)->position.at(j));
+        particles->at(i)->position.at(j) += velocities.at(i*NUM_OF_DIMENSIONS + j);
+      }
     }
 
     OMEGA = OMEGA * 0.9f;
 
-    for (int i = 0; i < NUM_OF_PARTICLES*NUM_OF_DIMENSIONS; i += NUM_OF_DIMENSIONS)
+    for (int i = 0; i < NUM_OF_PARTICLES; i++)
     {
-      for (int j = 0; j < NUM_OF_DIMENSIONS; j++)
-      {
-        tempParticle1[j] = positions[i + j];
-        tempParticle2[j] = pBests[i + j];
+      Particle* p = particles->at(i);
+      Particle* curPBest = pBest.at(i);
+
+      // Get the NCC of the current particle
```

Comment:
```suggestion
      // Get the NCC values of all particles in this iteration
```

Reply:
I would disagree with this change, the line below that comment is only getting the NCC value for the current particle as defined by the loop above. Each iteration, is when the `while` loop increments.

Reply:
kk - so we are still looping through the particle set to evaluate ncc in the given iteration- but updating the position and velocity via the particle vector (set of all particles?)

Would it make sense for the particle class to have it's velocity too?

Reply:
Yes, it would, since each velocity is unique to each particle. Nice catch, I will request your review once I add that.